Infrastructure tweaks to support error-measure based stepping

moxcodes opened the pull request:
## Proposed changes

Various small changes to time steppers, step choosers, and step controllers that preserve existing functionality, but allow the possibility of an iterative process to support step rejection and retry.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

moxcodes committed:
Add error_estimate_order function to `TimeStepper`s

moxcodes committed:
Permit different type for TimeStepper error measure

- necessary to support the error measure update as a `Variables` with a
different tag prefix

moxcodes committed:
Adjust step operations to permit repeated calls to `update_u`

This is needed for step-unwinding. This adjusts the interface by consistently
using the latest volume data in the history instead of using the current `u`
value in some steppers

moxcodes committed:
Add step acceptance parameter to `StepChooser`s

moxcodes commented:
Removed a small part that better fits in a different PR.

moxcodes commented:
I've dropped the step controller change, and added a fixup for the step chooser documentation.
For the moment, I've prefixed the `Separate timestepper history clean-up` with `drop`, as @wthrowe plans to submit a separate PR to perform a similar task in a way that will work better for dense output features; Once that is in, I'll rebase this PR and drop the commit associated with history clean-up from this one.

wthrowe commented on src/Time/TimeSteppers/AdamsBashforthN.hpp:
```diff
@@ -426,11 +425,17 @@ void AdamsBashforthN::update_u_impl(const gsl::not_null<UpdateVars*> u,
                                     const History<Vars, DerivVars>& history,
                                     const Delta& time_step,
                                     const size_t order) const noexcept {
-  ASSERT(history.size() <= order_,
+  ASSERT(history.size() <= order_ + 1,
          "Length of history (" << history.size() << ") "
-         << "should not exceed target order (" << order_ << ")");
+         << "should not exceed target order (" << order_ << ") by more than 1");
```
What is this change about?

moxcodes replied:
Ah, yeah good catch -- these couple of lines of the hunk should be dropped. It was for an earlier and less elegant attempt at the history cleanup change.

wthrowe commented on tests/Unit/Helpers/Time/TimeSteppers/TimeStepperTestUtils.cpp:
```diff
@@ -141,7 +150,9 @@ void integrate_test(const TimeStepper& stepper,
   for (uint64_t i = 0; i < num_steps; ++i) {
-    take_step(&time, &y, &history, stepper, rhs, step_size);
+    // last parameter: check that we can apply the stepper sometime in the
+    // middle of the evolution without messing anything up.
```
"apply the stepper twice", maybe?  Again later.

wthrowe reviewed (changes requested):
I'm having trouble quickly getting the history management working in a way compatible with both this and dense output, so let's go with this form so we're not blocking you and I'll change it for dense output later.

moxcodes commented:
Dropped the part of the hunk that shouldn't have been present, added a fixup for the comment change, and removed the `drop:` prefix from the commit about the history cleanup.

wthrowe commented:
Looks good.  Squash.

moxcodes commented:
Squashed.
Cheers!

nilsvu commented on src/Time/StepChoosers/StepChooser.hpp:
```diff
@@ -60,20 +60,25 @@ class StepChooser : public PUP::able {
+  /// a strictly smaller step, guided by the requested time step value.
   template <typename Metavariables, typename DbTags>
-  double desired_step(
+  std::pair<double, bool> desired_step(
```
Would it be reasonable to make the return value here a `std::optional<double>`, meaning either "do a step with this size" or "don't do a step"? So the question is: Will the "desired step" `double` be used also when the "acceptance" `bool` is `false`? If this is what you mean with "guided by the requested time step value", please clarify, e.g. by rewording to "... the desired step value returned by this function".

moxcodes replied:
The return value is used when the step is rejected to choose the step size when retrying the step; it is assumed that when the step is rejected the returned step size is strictly smaller than the previously tried step. I'll clarify the documentation.

nilsvu reviewed (commented):
LGTM, though I'm by no means an expert in time-stepping :)

moxcodes commented:
Squashed in documentation rewording.
Cheers!

nilsvu commented:
@wthrowe could you take another look at this?

wthrowe commented:
@moxcodes What do you want to do with respect to #2883?  Do you want to drop your history changes here or have me revert them when I'm ready to merge that?

wthrowe commented:
Also, @nilsleiffischer, please give me some time to respond to things before pinging me.  I find it annoying.

moxcodes commented:
@wthrowe Would it be too much hassle for you to split out the commit that does the history clean-up at the start of the step to a different PR? Then that could be a dependency of both this (with my history clean up change dropped) and your larger PR that makes the self-start changes
The part here that checks that the step can be re-applied depends on the history clean up changes... If I understand the changes in #2883 correctly, there's no change to the cleanup for substep methods, which means that those methods won't be able to be re-applied (because on the conclusion of the full step, the history will be cleaned up and unavailable for retries). I think that's fine with me for now, I can just remove the expectation of substep re-application from this PR as well -- were any of the features that you needed dependent on being able to repeat `update_u` in substep methods?

wthrowe commented:
I don't need the update_u changes.  I think I just took them to avoid some git conflict.  I do use "Add order function to `TimeStepper`s", but that's simple so I can just cherry-pick it.

You are correct that I haven't done the substep methods, but redoing a substep probably isn't useful anyway.  (And redoing a full step is a different challenge.)  Substep methods will need to be converted eventually for dense output.

Submitted as #2894.

moxcodes commented:
rebased following merge of #2894, and dropped the history cleanup parts.
Cheers!
